Enqueue the distribution of user keys to the proxy servers.

blueandgold opened the pull request:
- this PR covers the ability to enqueue background jobs
- enqueue the key distribution as a background job; I think we want to enqueue each server separately as a job so that it's easier for the workers to pickup; but if not please let me know of a better way
- add clock process to schedule job, i.e. what's the trigger to enqueue jobs 
- I have manually tested this for now on my heroku instance, but i think we can write a functional test later on
- heroku will not start workers automatically; does anyone know a better process to start the worker?
  https://devcenter.heroku.com/articles/procfile#scaling-a-process-type
  look for "other process types don’t start by default"

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/uproxy/ufo-management-server-flask/27)

<!-- Reviewable:end -->

blueandgold committed:
Job queue

blueandgold committed:
Add the worker.py

blueandgold commented:
As always @jpevarnek , please take a look here!

blueandgold committed:
Enqueue the key distribution job.

blueandgold committed:
add docstring

blueandgold committed:
update requirements.txt and clock.py docstring

blueandgold commented:
Updating this include the clock process so that we can schedule jobs, as well as actually enqueue the key distribution as a background job.

blueandgold committed:
add TODO

xwsxethan commented:
Looks good overall. I think the way you're enqueueing should be fine as long as it works. I'll look at the doc to see about firing off the process.

---

Reviewed 2 of 4 files at r1, 2 of 3 files at r2, 3 of 3 files at r3.
Review status: 6 of 7 files reviewed at latest revision, 7 unresolved discussions.

---

*[cat, line 0 \[r3\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27#-K9hb8walQU9U2s5dTQ9:-K9hb8wbYKztvMxj8uva:408334267) ([raw file](https://github.com/uproxy/ufo-management-server-flask/blob/c767f83d869100402f341fad5832d8cbe54fb021/cat#L0)):*
What is this and is it necessary?

---

*[clock.py, line 8 \[r3\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27#-K9hcu5zlvSLcjLJ3hxj:-K9hcu6-EyY0AzhfMBaL:526345524) ([raw file](https://github.com/uproxy/ufo-management-server-flask/blob/207121950248504b6768e5bd7c6201f7c7525c34/clock.py#L8)):*
Nit: If this is a function, then we shouldn't be importing it directly.

---

*[clock.py, line 11 \[r3\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27#-K9hcMigTT5qSACtm1yX:-K9hcMihydBqAZHaMaFK:534888480) ([raw file](https://github.com/uproxy/ufo-management-server-flask/blob/207121950248504b6768e5bd7c6201f7c7525c34/clock.py#L11)):*
NIt: Since this is essentially a module-level constant, it should probably be CAPS_WITH_UNDERSCORES.

Also, mocking this out can be very difficult as you saw in my readability review. Is there another way to do this with just calling the function in the decorator? If not, it isn't a big deal, just something to think about.

---

*[clock.py, line 15 \[r3\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27#-K9hbPWG6Ka1uiEg2bVX:-K9hbPWG6Ka1uiEg2bVY:351673237) ([raw file](https://github.com/uproxy/ufo-management-server-flask/blob/207121950248504b6768e5bd7c6201f7c7525c34/clock.py#L15)):*
Nit: spelling of distribute/distribution

---

*[clock.py, line 18 \[r3\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27#-K9hbUUyioS-dyQ0sI9G:-K9hbUUyioS-dyQ0sI9H:1606971693) ([raw file](https://github.com/uproxy/ufo-management-server-flask/blob/207121950248504b6768e5bd7c6201f7c7525c34/clock.py#L18)):*
What are the print statements for?

---

*[ufo/proxy_server.py, line 132 \[r3\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27#-K9hbd97lfY4LSjgdz8r:-K9hbd97lfY4LSjgdz8s:-287940553) ([raw file](https://github.com/uproxy/ufo-management-server-flask/blob/207121950248504b6768e5bd7c6201f7c7525c34/ufo/proxy_server.py#L132)):*
Nit: Add Returns block

---

*[worker.py, line 27 \[r3\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27#-K9hcAbmSw8GqL5-cJvJ:-K9hcAbmSw8GqL5-cJvK:234886777) ([raw file](https://github.com/uproxy/ufo-management-server-flask/blob/207121950248504b6768e5bd7c6201f7c7525c34/worker.py#L27)):*
NIt: Since these are essentially module-level constants, they should probably be CAPS_WITH_UNDERSCORES.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27#-:-K9hdFPRWM8aRU_pUxfp:958342872)*
<!-- Sent from Reviewable.io -->

blueandgold committed:
address ethan comments

blueandgold commented:
Thanks for the review.  PTAL!  Cool, thanks for looking into starting up the worker dyno.  This will also need to be done for the clock dyno as well.

heroku ps:scale worker=1
heroku ps:scale clock=1

---

Review status: 6 of 7 files reviewed at latest revision, 7 unresolved discussions.

---

*[clock.py, line 8 \[r3\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27#-K9hcu5zlvSLcjLJ3hxj:-K9hfFL1YA-hF6aF_Laj:-499146009) ([raw file](https://github.com/uproxy/ufo-management-server-flask/blob/207121950248504b6768e5bd7c6201f7c7525c34/clock.py#L8)):*
Done.

---

*[clock.py, line 11 \[r3\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27#-K9hcMigTT5qSACtm1yX:-K9hfUkw9g24ck-d3o1G:986847095) ([raw file](https://github.com/uproxy/ufo-management-server-flask/blob/207121950248504b6768e5bd7c6201f7c7525c34/clock.py#L11)):*
Done.  As far as using it in the decorator, that came straight from the documentation.  I think it's okay as I don't think there's anything interesting to actually test here on the unit level.  I would think that it's better to kick this up to the functional level?  But I am open to suggestions.

---

*[clock.py, line 15 \[r3\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27#-K9hbPWG6Ka1uiEg2bVX:-K9hgGrSoWO4IApkFlwF:-499146009) ([raw file](https://github.com/uproxy/ufo-management-server-flask/blob/207121950248504b6768e5bd7c6201f7c7525c34/clock.py#L15)):*
Done.

---

*[clock.py, line 18 \[r3\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27#-K9hbUUyioS-dyQ0sI9G:-K9hgL2bvRLPcENGNVKc:-500379254) ([raw file](https://github.com/uproxy/ufo-management-server-flask/blob/207121950248504b6768e5bd7c6201f7c7525c34/clock.py#L18)):*
Heroku logging actually only captures stdout or errors.  So, we should have a logger here that redirects to stdout, but I thought it's simpler here.  Adding a TOOD to address this separately.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27#-:-K9hgtwyrhFbBLoN0k16:-2144410571)*
<!-- Sent from Reviewable.io -->

xwsxethan commented:
Reviewed 1 of 1 files at r4, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[clock.py, line 11 \[r3\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27#-K9hcMigTT5qSACtm1yX:-K9hi3bjPS2yHpdjPYA3:1732047904) ([raw file](https://github.com/uproxy/ufo-management-server-flask/blob/207121950248504b6768e5bd7c6201f7c7525c34/clock.py#L11)):*
That's probably fine then.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27)*
<!-- Sent from Reviewable.io -->

blueandgold commented:
I am still in the process of testing the clock process (but out of time for now).  But the general thrust of this PR should still be good, which I appreciate the review.

jpevarnek commented:
Quick design question here: can you explain the decision for why we're using Redis here?  Thanks!

---

Reviewed 2 of 4 files at r1, 1 of 3 files at r2, 1 of 3 files at r3, 1 of 1 files at r4, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[cat, line 0 \[r5\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27#-K9mfevjoXwUID2cE1-k:-K9mfevjoXwUID2cE1-l:166874) ([raw file](https://github.com/uproxy/ufo-management-server-flask/blob/c767f83d869100402f341fad5832d8cbe54fb021/cat#L0)):*
?

---

*[clock.py, line 15 \[r5\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27#-K9mfsYdNkS7nF_kGWyA:-K9mfsYdNkS7nF_kGWyB:-1407819981) ([raw file](https://github.com/uproxy/ufo-management-server-flask/blob/fda425b975dbbb267c0ce93555d7044a5ea0baa0/clock.py#L15)):*
What is the printing for?

---

*[ufo/proxy_server.py, line 132 \[r5\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27#-K9mgT0j01CQGTm5hiEe:-K9mgT0kkEqn3pXiiv5y:-1679062744) ([raw file](https://github.com/uproxy/ufo-management-server-flask/blob/fda425b975dbbb267c0ce93555d7044a5ea0baa0/ufo/proxy_server.py#L132)):*
This is no longer a route.  The annotation should probably be removed and this should probably be moved to a different file.

---

*[worker.py, line 25 \[r5\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27#-K9mgiI3xBmmt6AcWyb5:-K9mgiI3xBmmt6AcWyb6:-420178204) ([raw file](https://github.com/uproxy/ufo-management-server-flask/blob/fda425b975dbbb267c0ce93555d7044a5ea0baa0/worker.py#L25)):*
This should be taken care of in `config` and set up to work anywhere in the application.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27#-:-K9mgDc0hhDsl9Fg-UDc:1944864130)*
<!-- Sent from Reviewable.io -->

blueandgold committed:
address review comments

blueandgold commented:
Good question: If you look at this link on task queue implementation, redis is what's listed as default for python.  I am agnostic on what we use, as long as it works and is well-supported.  Just like the apscheduler is the default for the clock process in the documentation.

https://devcenter.heroku.com/articles/background-jobs-queueing#implementation

---

Review status: 5 of 7 files reviewed at latest revision, 7 unresolved discussions.

---

*[clock.py, line 15 \[r5\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27#-K9mfsYdNkS7nF_kGWyA:-K9n6OGFTFsLa2HJwJlr:1109521429) ([raw file](https://github.com/uproxy/ufo-management-server-flask/blob/fda425b975dbbb267c0ce93555d7044a5ea0baa0/clock.py#L15)):*
I have not tested this yet, but I think that heroku logging facility will only pick up stdout and error.  So, I am using print for now and have a TODO to add a logger to redirect to stdout.  I do plan to test this to be 100% sure.

---

*[ufo/proxy_server.py, line 132 \[r3\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27#-K9hbd97lfY4LSjgdz8r:-K9n7_NFRYirphKaYrlm:-499146009) ([raw file](https://github.com/uproxy/ufo-management-server-flask/blob/207121950248504b6768e5bd7c6201f7c7525c34/ufo/proxy_server.py#L132)):*
Done.

---

*[ufo/proxy_server.py, line 132 \[r5\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27#-K9mgT0j01CQGTm5hiEe:-K9n7ljshaZ8bVN1nSNq:202718683) ([raw file](https://github.com/uproxy/ufo-management-server-flask/blob/fda425b975dbbb267c0ce93555d7044a5ea0baa0/ufo/proxy_server.py#L132)):*
Done, removing the setup annotation.  As far as moving to a different file, can I put a TODO here for a follow-up?  The main goal of this PR is to add the ability to use task/job queue.

---

*[worker.py, line 27 \[r3\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27#-K9hcAbmSw8GqL5-cJvJ:-K9n9Enb_AT2YuBuVZGx:1600161528) ([raw file](https://github.com/uproxy/ufo-management-server-flask/blob/207121950248504b6768e5bd7c6201f7c7525c34/worker.py#L27)):*
All done. 

---

*[worker.py, line 25 \[r5\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27#-K9mgiI3xBmmt6AcWyb5:-K9n8PfX2fA9lo_Dh_RG:610989145) ([raw file](https://github.com/uproxy/ufo-management-server-flask/blob/fda425b975dbbb267c0ce93555d7044a5ea0baa0/worker.py#L25)):*
I just want to double-check here, since the "worker" is considered to be a separate process from the "web" (see Procfile).  This might imply that we should not mix them all together in the same config?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27#-:-K9n50Eue4Wccv_VJHNT:-1850623349)*
<!-- Sent from Reviewable.io -->

blueandgold committed:
fix import reference

blueandgold committed:
fix import of proxyserver inside clock.py

blueandgold committed:
change import

blueandgold committed:
add comment on import

blueandgold committed:
remove the print statements as they don't seem to do anything either

blueandgold committed:
add a print for testing

blueandgold committed:
make the schedule shorter for testing

blueandgold committed:
add print for debugging

blueandgold committed:
add logging to clock

blueandgold committed:
add a proper print statement

blueandgold committed:
add a print to proxy server

blueandgold committed:
add more prints

blueandgold committed:
add print to see if job is enqueued

jpevarnek commented:
Reviewed 3 of 3 files at r6.
Review status: 5 of 7 files reviewed at latest revision, 8 unresolved discussions.

---

*[clock.py, line 15 \[r6\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27#-KA1HLprJGBR3s_9RnH2:-KA1HLprJGBR3s_9RnH3:-801120285) ([raw file](https://github.com/uproxy/ufo-management-server-flask/blob/593928634db8972d15c57a1e493754a28182096d/clock.py#L15)):*
Is this line intentional?

---

*[ufo/proxy_server.py, line 132 \[r5\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27#-K9mgT0j01CQGTm5hiEe:-KA1LzbhYU_z2H3KQlDX:-645011495) ([raw file](https://github.com/uproxy/ufo-management-server-flask/blob/fda425b975dbbb267c0ce93555d7044a5ea0baa0/ufo/proxy_server.py#L132)):*
I think we'd be fine with just moving this function, and anything else relevant, to the file where it's actually getting spawned.  I'm not quite sure what your reasoning for having it continue to live in the file for the routes is, if there is a compelling reason though, that's totally fine.  If it's a matter of not wanting to cut+paste the code though (and, though I may be misunderstanding, that's how it appears to me) I think it would be best to take care of it now.

---

*[worker.py, line 25 \[r5\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27#-K9mgiI3xBmmt6AcWyb5:-KA1MRbqkfP420eVufK7:307535431) ([raw file](https://github.com/uproxy/ufo-management-server-flask/blob/fda425b975dbbb267c0ce93555d7044a5ea0baa0/worker.py#L25)):*
Ah, yeah, if you want to have that separate, then we can have a totally separate way to handle it.

If we want to have it as a totally independent server processes though, let's try to move the code to totally separate directories then for clarity.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27)*
<!-- Sent from Reviewable.io -->

blueandgold committed:
add more print for output

blueandgold committed:
remove prints

blueandgold committed:
remove unused logging import

blueandgold commented:
@xwsxethan @jpevarnek PTAL: I have tested both the worker and clock processes.

blueandgold committed:
refactor key distribution to it's own module

blueandgold committed:
move key distributor import to __init__

blueandgold committed:
change schedule

blueandgold committed:
remove cat file

blueandgold commented:
Review status: 4 of 9 files reviewed at latest revision, 8 unresolved discussions.

---

*[clock.py, line 15 \[r5\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27#-K9mfsYdNkS7nF_kGWyA:-KA1fyaZ-siGdWHOYDte:447413573) ([raw file](https://github.com/uproxy/ufo-management-server-flask/blob/fda425b975dbbb267c0ce93555d7044a5ea0baa0/clock.py#L15)):*
This is now tested and confirmed.  I filed a bug to think about how to handle this.

---

*[clock.py, line 15 \[r6\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27#-KA1HLprJGBR3s_9RnH2:-KA1ftfYfo6faL1Fdpdz:-340781049) ([raw file](https://github.com/uproxy/ufo-management-server-flask/blob/593928634db8972d15c57a1e493754a28182096d/clock.py#L15)):*
Done. Removed

---

*[ufo/proxy_server.py, line 132 \[r5\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27#-K9mgT0j01CQGTm5hiEe:-KA6aLbi4EXSBi_W-mKI:1371796917) ([raw file](https://github.com/uproxy/ufo-management-server-flask/blob/fda425b975dbbb267c0ce93555d7044a5ea0baa0/ufo/proxy_server.py#L132)):*
Done.  I didn't mean not to do the refactoring.  I just didn't want to move things around too much in this PR, that would muddy the implementation of the job queue and clock processes here.  :)  So, this is now moved to a key_distributor.py and cleaned up the method names there so that they make sense in the context of a queue now.

---

*[cat, line  \[r5\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27#-K9mfevjoXwUID2cE1-k:-KA6bvfcxfpBghXqXJky:610327451) ([raw file](https://github.com/uproxy/ufo-management-server-flask/blob/c767f83d869100402f341fad5832d8cbe54fb021/cat#L)):*
Removed

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27)*
<!-- Sent from Reviewable.io -->

blueandgold commented:
Review status: 4 of 9 files reviewed at latest revision, 8 unresolved discussions.

---

*[cat, line  \[r3\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27#-K9hb8walQU9U2s5dTQ9:-KA6dJ2Bd_bEhNNd-d75:610327451) ([raw file](https://github.com/uproxy/ufo-management-server-flask/blob/c767f83d869100402f341fad5832d8cbe54fb021/cat#L)):*
Removed

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27)*
<!-- Sent from Reviewable.io -->

blueandgold commented:
Review status: 4 of 9 files reviewed at latest revision, 8 unresolved discussions.

---

*[worker.py, line 25 \[r5\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27#-K9mgiI3xBmmt6AcWyb5:-KA6ddk0r8279HyJSTyZ:1410624879) ([raw file](https://github.com/uproxy/ufo-management-server-flask/blob/fda425b975dbbb267c0ce93555d7044a5ea0baa0/worker.py#L25)):*
Hmmm, I tried to move worker.py to it's own package, but internally, redis queue is expecting it to be at the top-level.
`rq.exceptions.UnpickleError: (u'Could not unpickle', ImportError('No module named worker',))`

I don't quite see a way to configure where redis queue is looking for this, so can we leave this here?  I think overall, having worker.py living at the top-level is not all that bad.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27)*
<!-- Sent from Reviewable.io -->

blueandgold committed:
add unit test for make string

blueandgold commented:
@xwsxethan @jpevarnek PTAL here.  I did the refactor, as well as adding some basic unit test.

xwsxethan commented:
My apologies for the delay. I didn't realize you wanted me to look until I was scrolling through this PR again just to check the status. Comments are mostly nits about imports being absolute vs relative and indentation with a few questions about private vs public methods. Looks good overall structurally and functionally to me though. :+1: after fixes.

---

Reviewed 1 of 3 files at r6, 8 of 8 files at r8.
Review status: all files reviewed at latest revision, 22 unresolved discussions.

---

*[ufo/key_distributor.py, line 3 \[r8\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27#-KABD3lwvysct5UUZCF8:-KABD3lwvysct5UUZCF9:-469614885) ([raw file](https://github.com/uproxy/ufo-management-server-flask/blob/83daf77a92bfb485a0f16141eed0c9d9a7039190/ufo/key_distributor.py#L3)):*
Nit: I have gone back and forth with this, but I think the style guide wants absolute imports, so should probably be from ufo import models.

---

*[ufo/key_distributor.py, line 14 \[r8\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27#-KABDQ6DrikgqOB0ZEeA:-KABDQ6ETJygBs4jRd2U:867380369) ([raw file](https://github.com/uproxy/ufo-management-server-flask/blob/83daf77a92bfb485a0f16141eed0c9d9a7039190/ufo/key_distributor.py#L14)):*
Nit: extra indent space should not be there.

---

*[ufo/key_distributor.py, line 17 \[r8\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27#-KABDWPfHatcHpYl_5YQ:-KABDWPg5Oq4DNg1Ikyc:1460580902) ([raw file](https://github.com/uproxy/ufo-management-server-flask/blob/83daf77a92bfb485a0f16141eed0c9d9a7039190/ufo/key_distributor.py#L17)):*
Nit: indent

---

*[ufo/key_distributor.py, line 32 \[r8\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27#-KABDZkL2t16F_DQ1EPw:-KABDZkL2t16F_DQ1EPx:1460580902) ([raw file](https://github.com/uproxy/ufo-management-server-flask/blob/83daf77a92bfb485a0f16141eed0c9d9a7039190/ufo/key_distributor.py#L32)):*
Nit: indent

---

*[ufo/key_distributor.py, line 34 \[r8\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27#-KABD__yxVqu6vRszaZp:-KABD__yxVqu6vRszaZq:1460580902) ([raw file](https://github.com/uproxy/ufo-management-server-flask/blob/83daf77a92bfb485a0f16141eed0c9d9a7039190/ufo/key_distributor.py#L34)):*
Nit: indent

---

*[ufo/key_distributor.py, line 37 \[r8\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27#-KABDhnfNcYzsiL7jLY6:-KABDhnfNcYzsiL7jLY7:1460580902) ([raw file](https://github.com/uproxy/ufo-management-server-flask/blob/83daf77a92bfb485a0f16141eed0c9d9a7039190/ufo/key_distributor.py#L37)):*
Nit: indent

---

*[ufo/key_distributor.py, line 44 \[r8\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27#-KABDnX9NcWrh0g1IfLv:-KABDnXAdH45kZja6Dk6:1460580902) ([raw file](https://github.com/uproxy/ufo-management-server-flask/blob/83daf77a92bfb485a0f16141eed0c9d9a7039190/ufo/key_distributor.py#L44)):*
Nit: indent

---

*[ufo/key_distributor.py, line 46 \[r8\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27#-KABDqIqs3NHEgFJA3jp:-KABDqIqs3NHEgFJA3jq:1460580902) ([raw file](https://github.com/uproxy/ufo-management-server-flask/blob/83daf77a92bfb485a0f16141eed0c9d9a7039190/ufo/key_distributor.py#L46)):*
Nit: indent

---

*[ufo/key_distributor.py, line 48 \[r8\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27#-KABDofPbx-3I36q7fKc:-KABDofPbx-3I36q7fKd:1460580902) ([raw file](https://github.com/uproxy/ufo-management-server-flask/blob/83daf77a92bfb485a0f16141eed0c9d9a7039190/ufo/key_distributor.py#L48)):*
Nit: indent

---

*[ufo/key_distributor.py, line 55 \[r8\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27#-KABENzrEjv2q2aojc9u:-KABENzrEjv2q2aojc9v:-188597131) ([raw file](https://github.com/uproxy/ufo-management-server-flask/blob/83daf77a92bfb485a0f16141eed0c9d9a7039190/ufo/key_distributor.py#L55)):*
Since we're passing around the distribute key function, should it just be public? I know it doesn't technically matter in Python, but just curious.

---

*[ufo/key_distributor_test.py, line 5 \[r8\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27#-KABDw14qxr9YhxULzP3:-KABDw14qxr9YhxULzP4:-1108128404) ([raw file](https://github.com/uproxy/ufo-management-server-flask/blob/83daf77a92bfb485a0f16141eed0c9d9a7039190/ufo/key_distributor_test.py#L5)):*
Nit: absolute imports instead of relative as mentioned above.

---

*[ufo/key_distributor_test.py, line 23 \[r8\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27#-KABE7TPi4e8Buu-5mQF:-KABE7TPi4e8Buu-5mQG:1460580902) ([raw file](https://github.com/uproxy/ufo-management-server-flask/blob/83daf77a92bfb485a0f16141eed0c9d9a7039190/ufo/key_distributor_test.py#L23)):*
Nit: indent

---

*[ufo/key_distributor_test.py, line 24 \[r8\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27#-KABEot11XmQNmTW7lGw:-KABEot11XmQNmTW7lGx:355109532) ([raw file](https://github.com/uproxy/ufo-management-server-flask/blob/83daf77a92bfb485a0f16141eed0c9d9a7039190/ufo/key_distributor_test.py#L24)):*
Since we're pretty much only asserting on the key string directly in both cases here, you should probably just make _make_key_string a public function or mock out the queue.enqueue call and assert against what is passed that way.

---

*[ufo/key_distributor_test.py, line 40 \[r8\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27#-KABEB126n7JOkj5CVGc:-KABEB13JV5m8F9-_yTp:1460580902) ([raw file](https://github.com/uproxy/ufo-management-server-flask/blob/83daf77a92bfb485a0f16141eed0c9d9a7039190/ufo/key_distributor_test.py#L40)):*
Nit: indent

---

*[ufo/proxy_server.py, line 6 \[r8\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27#-KABFERl-wcywjt4hDJw:-KABFERmyPYoRTK1dOZ0:-2146998730) ([raw file](https://github.com/uproxy/ufo-management-server-flask/blob/83daf77a92bfb485a0f16141eed0c9d9a7039190/ufo/proxy_server.py#L6)):*
Nit: absolute imports instead of relative

---

*[ufo/user_test.py, line 24 \[r8\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27#-KABFvv6zWNWrs9OYxQP:-KABFvv6zWNWrs9OYxQQ:1952075673) ([raw file](https://github.com/uproxy/ufo-management-server-flask/blob/83daf77a92bfb485a0f16141eed0c9d9a7039190/ufo/user_test.py#L24)):*
If this list of fake emails and names is useful for other tests, we can move it into the base_test. I can take a look at this myself based on what is in the database and other tests, just curious on your thoughts here.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27#-:-KABGL_MaBAni4vdnIBn:-1824146974)*
<!-- Sent from Reviewable.io -->

blueandgold committed:
test import

blueandgold committed:
address ethan's comments

blueandgold commented:
Thanks for the review!  All comments are addressed.  Will give Jonathan a chance to look as well.

---

Review status: 8 of 11 files reviewed at latest revision, 21 unresolved discussions.

---

*[ufo/key_distributor.py, line 3 \[r8\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27#-KABD3lwvysct5UUZCF8:-KABc6jaHs_clXmwlXxq:-499146009) ([raw file](https://github.com/uproxy/ufo-management-server-flask/blob/83daf77a92bfb485a0f16141eed0c9d9a7039190/ufo/key_distributor.py#L3)):*
Done.

---

*[ufo/key_distributor.py, line 14 \[r8\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27#-KABDQ6DrikgqOB0ZEeA:-KABeLBZ4OgUErpewWvR:-499146009) ([raw file](https://github.com/uproxy/ufo-management-server-flask/blob/83daf77a92bfb485a0f16141eed0c9d9a7039190/ufo/key_distributor.py#L14)):*
Done.

---

*[ufo/key_distributor.py, line 17 \[r8\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27#-KABDWPfHatcHpYl_5YQ:-KABeLbYmd0r4Au6l3bb:-499146009) ([raw file](https://github.com/uproxy/ufo-management-server-flask/blob/83daf77a92bfb485a0f16141eed0c9d9a7039190/ufo/key_distributor.py#L17)):*
Done.

---

*[ufo/key_distributor.py, line 32 \[r8\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27#-KABDZkL2t16F_DQ1EPw:-KABeNmtlPy00t9XPEHx:-499146009) ([raw file](https://github.com/uproxy/ufo-management-server-flask/blob/83daf77a92bfb485a0f16141eed0c9d9a7039190/ufo/key_distributor.py#L32)):*
Done.

---

*[ufo/key_distributor.py, line 34 \[r8\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27#-KABD__yxVqu6vRszaZp:-KABeQP7_wYufg6eezum:-499146009) ([raw file](https://github.com/uproxy/ufo-management-server-flask/blob/83daf77a92bfb485a0f16141eed0c9d9a7039190/ufo/key_distributor.py#L34)):*
Done.

---

*[ufo/key_distributor.py, line 37 \[r8\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27#-KABDhnfNcYzsiL7jLY6:-KABeS17NCIMAIKDymPx:-499146009) ([raw file](https://github.com/uproxy/ufo-management-server-flask/blob/83daf77a92bfb485a0f16141eed0c9d9a7039190/ufo/key_distributor.py#L37)):*
Done.

---

*[ufo/key_distributor.py, line 44 \[r8\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27#-KABDnX9NcWrh0g1IfLv:-KABeThBkrq2pc_KaIHs:-499146009) ([raw file](https://github.com/uproxy/ufo-management-server-flask/blob/83daf77a92bfb485a0f16141eed0c9d9a7039190/ufo/key_distributor.py#L44)):*
Done.

---

*[ufo/key_distributor.py, line 46 \[r8\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27#-KABDqIqs3NHEgFJA3jp:-KABeVHoWMwXxuFW0R4G:-499146009) ([raw file](https://github.com/uproxy/ufo-management-server-flask/blob/83daf77a92bfb485a0f16141eed0c9d9a7039190/ufo/key_distributor.py#L46)):*
Done.

---

*[ufo/key_distributor.py, line 48 \[r8\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27#-KABDofPbx-3I36q7fKc:-KABeWeZmqylOyfPqWFf:-499146009) ([raw file](https://github.com/uproxy/ufo-management-server-flask/blob/83daf77a92bfb485a0f16141eed0c9d9a7039190/ufo/key_distributor.py#L48)):*
Done.

---

*[ufo/key_distributor.py, line 55 \[r8\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27#-KABENzrEjv2q2aojc9u:-KABeb4NMwPx0WD0jxxm:-1587701861) ([raw file](https://github.com/uproxy/ufo-management-server-flask/blob/83daf77a92bfb485a0f16141eed0c9d9a7039190/ufo/key_distributor.py#L55)):*
The distribute key function here is being passed around internally (i.e. within this class).  If the function is access outside of this class, then yes it should be public.  But if that happens then yes, it should be public; so I'm going to leave it private for now.

i.e. we are not doing this (for now), 
KeyDistributor().distribute_key

---

*[ufo/key_distributor_test.py, line 5 \[r8\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27#-KABDw14qxr9YhxULzP3:-KABfNtTppcygBuWndLB:-499146009) ([raw file](https://github.com/uproxy/ufo-management-server-flask/blob/83daf77a92bfb485a0f16141eed0c9d9a7039190/ufo/key_distributor_test.py#L5)):*
Done.

---

*[ufo/key_distributor_test.py, line 23 \[r8\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27#-KABE7TPi4e8Buu-5mQF:-KABfQczTqsAafbnoj5Q:-499146009) ([raw file](https://github.com/uproxy/ufo-management-server-flask/blob/83daf77a92bfb485a0f16141eed0c9d9a7039190/ufo/key_distributor_test.py#L23)):*
Done.

---

*[ufo/key_distributor_test.py, line 24 \[r8\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27#-KABEot11XmQNmTW7lGw:-KABghLxbA-97AOTCx5z:-873126515) ([raw file](https://github.com/uproxy/ufo-management-server-flask/blob/83daf77a92bfb485a0f16141eed0c9d9a7039190/ufo/key_distributor_test.py#L24)):*
Done. Making this public here.

---

*[ufo/key_distributor_test.py, line 40 \[r8\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27#-KABEB126n7JOkj5CVGc:-KABgmgGGfyjAKGTKsDV:-499146009) ([raw file](https://github.com/uproxy/ufo-management-server-flask/blob/83daf77a92bfb485a0f16141eed0c9d9a7039190/ufo/key_distributor_test.py#L40)):*
Done.

---

*[ufo/proxy_server.py, line 6 \[r8\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27#-KABFERl-wcywjt4hDJw:-KABg_UJkCeQRMQkwtZo:-499146009) ([raw file](https://github.com/uproxy/ufo-management-server-flask/blob/83daf77a92bfb485a0f16141eed0c9d9a7039190/ufo/proxy_server.py#L6)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27#-:-K9nHkakOnIChP0YCFFU:-276939880)*
<!-- Sent from Reviewable.io -->

jpevarnek commented:
Reviewed 5 of 8 files at r8, 3 of 3 files at r9.
Review status: all files reviewed at latest revision, 18 unresolved discussions.

---

*[ufo/key_distributor.py, line 32 \[r9\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27#-KAHpP0SS2oNFM5Ipz42:-KAHpP0Url5ZArnilGom:-1486478096) ([raw file](https://github.com/uproxy/ufo-management-server-flask/blob/2ff80f9eea256b91d8471008ab3aa4d7ebb8a4a2/ufo/key_distributor.py#L32)):*
Is there a reason we're using the variable names instead of just the string literals here?  For me at least, `space` seems no more legible than `' '`

---

*[worker.py, line 25 \[r5\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27#-K9mgiI3xBmmt6AcWyb5:-KAHqRlzViSHbv_nrXdC:-763728348) ([raw file](https://github.com/uproxy/ufo-management-server-flask/blob/fda425b975dbbb267c0ce93555d7044a5ea0baa0/worker.py#L25)):*
Sure, good for now.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27)*
<!-- Sent from Reviewable.io -->

blueandgold committed:
address jonathan's comments

blueandgold commented:
Thanks for the review.  PTAL!

---

Review status: 10 of 11 files reviewed at latest revision, 18 unresolved discussions, some commit checks failed.

---

*[ufo/key_distributor.py, line 32 \[r9\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27#-KAHpP0SS2oNFM5Ipz42:-KAHzpylyTX0GTaSNP6H:-508449480) ([raw file](https://github.com/uproxy/ufo-management-server-flask/blob/2ff80f9eea256b91d8471008ab3aa4d7ebb8a4a2/ufo/key_distributor.py#L32)):*
No, I didn't write this originally.  But changing this now. 

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27#-:-KAI-WQ5rylf0prPm0kz:-433514316)*
<!-- Sent from Reviewable.io -->

xwsxethan commented:
Reviewed 2 of 3 files at r9, 1 of 1 files at r10.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[ufo/key_distributor.py, line 32 \[r9\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27#-KAHpP0SS2oNFM5Ipz42:-KALBRTQQwnayXNYMaHy:481339481) ([raw file](https://github.com/uproxy/ufo-management-server-flask/blob/2ff80f9eea256b91d8471008ab3aa4d7ebb8a4a2/ufo/key_distributor.py#L32)):*
I wrote it originally, and I was trying to clean it up from having so many literals with plusses. To be honest, the best way would be with a stringbuilder, such as %s for the variables and % or format to build the string. The space and endline variables in that case could just be part of the literal that is then formatted. Also, note you only changed one space variable over and missed the second one in the last commit.

Example since I realize this is confusing as I write it:
user_string = '%s %s %s\n'.format(ssh_starting_portion, user.public_key, user.email)
or
user_string = '{} {} {}\n' % (ssh_starting_portion, user.public_key, user.email)

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27)*
<!-- Sent from Reviewable.io -->

blueandgold committed:
fix the other space

blueandgold commented:
Review status: 10 of 11 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[ufo/key_distributor.py, line 32 \[r9\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27#-KAHpP0SS2oNFM5Ipz42:-KALTi7JAgHjeDWPVwfR:1495276556) ([raw file](https://github.com/uproxy/ufo-management-server-flask/blob/2ff80f9eea256b91d8471008ab3aa4d7ebb8a4a2/ufo/key_distributor.py#L32)):*
Done, fixed the second space.  Surprisingly, despite that strings are immutable, stringbuilder or string formatting has no practical advantages over concatenation.

http://stackoverflow.com/a/376508/2830207

Personally, I much rather perfer concatenation for being easier to read.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/27)*
<!-- Sent from Reviewable.io -->

blueandgold committed:
reference public method

blueandgold committed:
merge conflict
